Show first-party cookie statistics in tracker-radar

sammacbeth opened the pull request:
https://app.asana.com/0/481882893211075/1199637690641235/f

Adds stats on first party cookies set by 3rd party scripts to tracker-radar data. This is added per resource as follows:
```js
"resources": [
    {
      "rule": "google-analytics\\.com\\/analytics\\.js",
      "firstPartyCookies": {
        "_ga": {
          "ttl": 63072001, // median expiry of this cookie, in seconds from set time.
          "length": 26, // mean string length of the cookie value
          "prevalence": 0.23242023108118448, // proportion of sites in the crawl where this cookie was set by this script
          "uniqueness": 1 // number of distinct values seen divided by times set
        },
    ...
```

We also detect when the cookie value was sent in another 3rd party request, and include the frequency of this:
```js
"resources": [
...
   {
      "firstPartyCookiesSent": {
        "_ga": 0.03884765475552835, // proportion of sites in the crawl where the value of this 1st party cookie was sent to this endpoint
        "G_ENABLED_IDPS": 0.0005126635549774275
      },
```

This PR is based on top of #26 so please merge that before merging this one.

sammacbeth committed:
Add tests for firstPartyCookies structure

sammacbeth committed:
Parsing document cookies per site.

sammacbeth committed:
Process first party cookies.

sammacbeth committed:
Add firstPartyCookiesSent stat.

sammacbeth committed:
Use prevalence

sammacbeth committed:
Add firstPartyCookiesSent to resource

sammacbeth committed:
Extract cookie length cutoff for first party cookies

sammacbeth committed:
Fix 3rd partyness filtering of document cookies for IPs

sammacbeth committed:
Set a threshold proportion of pages for a sent cookie to be included in commonrequests

sammacbeth committed:
Deduplicate document cookies with the same name.

sammacbeth committed:
Test that cookies set by local scripts are not included in results

sammacbeth committed:
Test cookie parsing

kdzwinel commented on src/trackers/build-trackers.js:
```diff
@@ -37,7 +37,7 @@ for (const key in newData.requests) {
     const rule = new Rule(newTrackerData, newData.stats.sites)
 
-    if (!(rule.cookies || rule.fingerprinting)) {
+    if (!(rule.cookies || rule.fingerprinting || Object.keys(rule.firstPartyCookiesSent).length > 0)) {
```
So now if we see a domain that collects first party cookies, w/o setting any cookies itself and w/o touching any APIs that would classify it as a fingerprinter, we put that domain into Tracker Radar Dataset, correct?

@jdorweiler what about the downstream? When we discussed relaxing those rules you mentioned that we need to tighten them up on the other end not to affect the blocklist.

sammacbeth replied:
Correct. Without this some of the examples we've discussed would not end up in the final dataset, for example `analytics.google.com` and `t.co`. I was assuming that by not changing the `cookies` and `fingerprinting` values this doesn't yet affect downstream lists.

kdzwinel replied:
When discussing https://app.asana.com/0/1128329641547046/1199881153638320/f , which is somehow similar, Jason suggested that we will need downstream changes. @jdorweiler WDYT?

kdzwinel replied:
I played with the blocklist building code - it will filter everything that's fp=0 & cookie=0 out, so we are good here 👍

jdorweiler commented on src/trackers/classes/request.js:
```diff
@@ -2,6 +2,8 @@ const shared = require('./../helpers/sharedData.js')
 const ParsedUrl = require('./../helpers/parseUrl.js')
 const getOwner = require('./../helpers/getOwner.js')
 
+const COOKIE_LENGTH_CUTOFF = 5
```
Why did you pick 5?

sammacbeth replied:
This is fairly arbitrary - the assumption is that values shorter than 5 don't have enough entropy to be of interest for tracking. We may want to come up with a better way to pick this value, but I would considered it out of scope for this PR.

kdzwinel commented on src/trackers/classes/request.js:
```diff
@@ -16,6 +18,18 @@ class Request {
         this.responseHash = reqData.responseBodyHash
         this.nameservers = []
+        this.firstPartyCookies = site.documentCookies
+            .filter(cookie => cookie.source === reqData.url && (!cookie.domain || cookie.domain.endsWith(site.domain)) && cookie.value && cookie.value.length > COOKIE_LENGTH_CUTOFF)
```
hm, this is a bit hard to read so I'd break it down if possible. Two thoughts:

> cookie.domain.endsWith(site.domain)

If cookie domain is "hub.com" and site is "github.com" those would incorrectly match, right?

> cookie.value.length > COOKIE_LENGTH_CUTOFF

Isn't filtering it out here messing up the `cookie.uniqueness = cookie.values.size / cookie.pages` stat?

sammacbeth replied:
> If cookie domain is "hub.com" and site is "github.com" those would incorrectly match, right?

Hmm... this condition kind of got built up as I worked on it, and is now a bit confusing. I'll split into a separate function with tests to make sure it works correctly in cases like you outlined.

> > cookie.value.length > COOKIE_LENGTH_CUTOFF
> 
> Isn't filtering it out here messing up the `cookie.uniqueness = cookie.values.size / cookie.pages` stat?

I assume that it is unlikely a script would used both long and short values for an identifier (which is what this analysis is concerned with). The filtering is intended to cut out less interesting cookies early on, to reduce the about of data to process, and the number of irrelevant cookies in the final results. I would also be ok with removing it though and seeing what this looks like.

sammacbeth replied:
I updated this filter condition to use a function for the 1st party check and added tests for it.

kdzwinel replied:
> I assume that it is unlikely a script would used both long and short values for an identifier

If it's not a big deal to check this I think we should.

sammacbeth replied:
Yep, I already removed this (https://github.com/duckduckgo/tracker-radar-detector/pull/28/commits/913c7c1a9ed76980bc2ae96b173a1e7303b09a03) and now doing a test run to ensure that this doesn't cause any issues

kdzwinel commented on src/trackers/classes/request.js:
```diff
@@ -16,6 +18,18 @@ class Request {
+            .filter(cookie => cookie.source === reqData.url && (!cookie.domain || cookie.domain.endsWith(site.domain)) && cookie.value && cookie.value.length > COOKIE_LENGTH_CUTOFF)
+        this.firstPartyCookiesSent = site.documentCookies
+            .filter(cookie => {
+                // only consider cookies 5 or more characters long
```
```suggestion
                // only consider cookies 6 or more characters long
```

kdzwinel commented on src/trackers/classes/request.js:
```diff
@@ -16,6 +18,18 @@ class Request {
+                if (cookie.value && cookie.value.length > COOKIE_LENGTH_CUTOFF) {
+                    // for longer cookie values, exclude the first 6 characters (GA cookies only send the suffix)
+                    const cookieSuffix = cookie.value.length > 10 ? cookie.value.slice(6) : cookie.value
+                    return this.url.indexOf(cookieSuffix) !== -1
```
ohhhh, ok we are looking for cookie value in the url of all request, got it. Might be worth adding a comment TBH. Couple of thoughts here:

1. looking for a cookie value in the whole url seems like something that can produce false positives, maybe it'd be better to limit the search to path & query of an url?
2. can we limit this GA-specific mitigation to GA only? e.g. only exclude those 6 chars if cookie name is `_ga`? For other cookies this may create false positives.

sammacbeth replied:
1. True, I purposely ignored that issue here, because to do this properly it would be nice to have the URL parsing changes I proposed (https://app.asana.com/0/1128329641547046/1199565593892363/f).
 2. Sure, that makes sense.

sammacbeth replied:
I've added a commit to:
 1. Only look on the url path and search params for the value
 2. Make specific handling for cookies (e.g. `_ga`) pluggable
 3. Test these things.

kdzwinel commented on src/trackers/classes/site.js:
```diff
@@ -31,6 +35,22 @@ class Site {
+
+        this.documentCookies = Object.values(siteData.data.apis.savedCalls
+            .filter(call => {
+                if (!isSavedCookieSetterCall(call) || !call.source.startsWith('http')) {
```
`!call.source.startsWith('http')` - hm, when does this happen?  Are there any legitimate cases or just misattributions in the crawl data?

sammacbeth replied:
I saw some cases when `call.source == 'about:blank'` in the data.

kdzwinel replied:
> about:blank

Ah, gotcha. AFAIK this means that the main document was the source (inline script in the main doc), but I'd need to double check.

kdzwinel commented on src/trackers/classes/site.js:
```diff
@@ -31,6 +35,22 @@ class Site {
 
         this.analyzeRequest = _analyzeRequest.bind(this)
+
+        this.documentCookies = Object.values(siteData.data.apis.savedCalls
```
I think we need to name this better or add a comment. From the name I thought those are all cookies (like in `document.cookies`), but those are only 1p cookies set by 3p scripts, correct? Maybe `thirdPartyJSCookies`?

kdzwinel commented on src/trackers/classes/site.js:
```diff
@@ -31,6 +35,22 @@ class Site {
+                    return false
+                }
+                const sourceHost = tldts.parse(call.source)
+                return sourceHost.isIp ? sourceHost.host !== this.host : sourceHost.domain !== this.domain
```
hmmm, can you please explain why we have to distinguish between those two cases? In case of IPs domain is not set?

sammacbeth replied:
When tldts parses an IP the `domain` is `null`, so using the domain check would mark any other IP as the same origin. When `sourceHost` is not an IP, then we use `domain` in order to compare eTLD+1.

(Source https://github.com/remusao/tldts/blob/master/packages/tldts/README.md)

kdzwinel commented on src/trackers/helpers/cookies.js:
```diff
@@ -0,0 +1,47 @@
+ * @param {Number} setAtTs - timestamp in ms when the cookie was set
+ * @returns TTL of this cookie in seconds from setAtTs.
+ */
+function calculateCookieTtl(cookie, setAtTs) {
```
what will happen with a session cookie here?

sammacbeth replied:
Good point. I added a test to cover this - session cookies will have a ttl of `NaN`, which I think actually makes sense. This will differentiate session cookies from any cookie with an expiry.

The interesting feature of using `NaN`, is that when we calculate the median at the end, if over half of the cookies are session cookies, we'll pick `NaN` as the ttl, showing that these are session cookies in the data.

kdzwinel replied:
One edge case is when there are two middle values and one of them is NaN and the other one is a proper number. But yeah, I think that NaN in the general case is better than using 0 or infinity here.

kdzwinel reviewed (commented):
Nice work! Can't wait to see that data in the Tracker Radar 🤩

I left a few questions.

sammacbeth committed:
Update src/trackers/classes/request.js

Co-authored-by: Konrad Dzwinel <kdzwinel@gmail.com>

sammacbeth committed:
Test cookie TTL.

sammacbeth committed:
Clean up first party cookie check and add tests

sammacbeth committed:
Remove length cutoff for per request 1st party cookies.

sammacbeth committed:
Improve detection of cookie values in URLs

Implements a pluggable system for creating special handling for specific cookies

sammacbeth committed:
Rename site.documentCookies -> thirdPartyJSCookies

kdzwinel commented on src/trackers/classes/commonRequest.js:
```diff
@@ -106,6 +136,25 @@ function _finalize (request, totalSites) {
+
+    Object.values(request.firstPartyCookies).forEach(cookie => {
+        cookie.length = Math.round(cookie.lengthSum / cookie.pages)
+        cookie.ttl = cookie.ttl.sort()[Math.floor(cookie.ttl.length / 2)]
```
I don't think it's a big deal, but maybe it makes sense to use a "full version" of a median instead (although that introduces this NaN + Number edge-case that I mentioned):

```js
function median(values) {
    if (values.length === 0) return 0;

    values.sort((a, b) => a - b);

    const half = Math.floor(values.length / 2);

    if (values.length % 2)
        return values[half];

    return (values[half - 1] + values[half]) / 2.0;
}
```

sammacbeth replied:
It seems we already use `mathjs` which includes a `median` implementation which is likely robust. We could use that: https://github.com/duckduckgo/tracker-radar-detector/blob/4bd5caa00a36fa4ffa6dcc230f07b449c0b6090b/src/trackers/classes/crawl.js#L2

kdzwinel replied:
Yeah, even better 👍

sammacbeth replied:
I quickly tested the mathjs median function, and it seems it will always return `NaN` if 1 or more `NaN` elements exists in the list. My implementation still works for this case.

kdzwinel commented on src/trackers/classes/request.js:
```diff
@@ -16,6 +19,17 @@ class Request {
+                // only consider cookies 6 or more characters long
+                return cookie.value &&
+                    cookie.value.length > COOKIE_LENGTH_CUTOFF &&
+                    isCookieValueInUrl(cookie, ParsedUrl.parse(reqData.url)) // TODO fix with url parsing update
```
is this TODO still needed?

sammacbeth replied:
Yes. I'll tackle this as LHF before the end of the week. The issue here is we have to parse the URL twice, as there is a `ParsedUrl` type as well as the standard `URL` being used.

kdzwinel commented on src/trackers/helpers/cookies.js:
```diff
@@ -0,0 +1,90 @@
+}, {
+    canHandle: () => true,
+    getValueSearchString: ({value}) => value,
+}]
```
Very neat 👌

kdzwinel reviewed (commented):
Thanks for those changes! I left couple more questions, but overall this LGTM.

sammacbeth commented:
Note, please squash when merging to tidy up the commit history :)

sammacbeth committed:
Fixes after url-parsing changes

jdorweiler commented:
Looks good, thanks. Can you make sure to update the Tracker Radar documentation when we do the next update?
